### ISSUE TITLE
error: undefined symbol: __isoc23_strtol

### ISSUE THREAD

**Issue by Aarqw12** [ClangBuiltLinux/linux#1929]
OS : Clear linux 39890

#library version of clear linux 

ldd --version : 2.38

LLVM used : LLVM 17.0.0-rc3

kernel source : https://android.googlesource.com/kernel/common/+/refs/heads/android14-6.1

command compile : tools/bazel run --config=fast //private/google-modules/soc/gs:slider_dist

error : 

```
ld.lld: error: undefined symbol: __isoc23_strtol
>>> referenced by stdlib.h:483 (/usr/include/stdlib.h:483)
>>>               libsubcmd-in.o:(pretty_print_string_list) in archive /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a
>>> referenced by stdlib.h:483 (/usr/include/stdlib.h:483)
>>>               libsubcmd-in.o:(pretty_print_string_list) in archive /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a
>>> referenced by stdlib.h:483 (/usr/include/stdlib.h:483)
>>>               libsubcmd-in.o:(pager_get_columns) in archive /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a
>>> referenced 2 more times

ld.lld: error: undefined symbol: __isoc23_strtoul
>>> referenced by parse-options.c:314 (/home/lcockx/Documents/pixel-6.1exp/common/tools/lib/subcmd/parse-options.c:314)
>>>               libsubcmd-in.o:(get_value) in archive /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a

ld.lld: error: undefined symbol: __isoc23_strtoull
>>> referenced by parse-options.c:332 (/home/lcockx/Documents/pixel-6.1exp/common/tools/lib/subcmd/parse-options.c:332)
>>>               libsubcmd-in.o:(get_value) in archive /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [Makefile:75: /home/lcockx/Documents/pixel-6.1exp/out/cache/020cfd9c/common/tools/bpf/resolve_btfids//resolve_btfids] Error 1
make[2]: *** [Makefile:76: bpf/resolve_btfids] Error 2
make[1]: *** [/home/lcockx/Documents/pixel-6.1exp/common/Makefile:1481: tools/bpf/resolve_btfids] Error 2
```

**Comment by nickdesaulniers** [comment]
I don't see explicit calls to `__isoc23_strtol`, `__isoc23_strtoul`, or `__isoc23_strtoull` in `pretty_print_string_list` defined in tools/lib/subcmd/help.c.

I suspect that they are ISO C23 string to long, unsigned long, and unsigned long long helpers for the calls to printf.

I wonder, are you or is your toolchain forcing `-std=c23` or `-std=c2x` for `HOSTCFLAGS` perhaps?  Then your libc (glibc perhaps) doesn't yet have support for those helper functions?

Though I'd expect the glibc headers to not have referenced those functions at all unless they were available to link against.

tools/lib/subcmd/Makefile:31:
```
CFLAGS := -ggdb3 -Wall -Wextra -std=gnu99 -fPIC
```
so this code should be built as `-std=gnu99`.  I highly suspect your build is doing something to override that.

**Comment by nathanchance** [comment]
I know that Clear Linux has system-wide CFLAGS for optimization purposes and while Kbuild should be relatively immune from that due to `KCFLAGS`, it is possible that Android's build system picks those up?

**Comment by nickdesaulniers** [comment]
Feedback from #glibc on IRC

> Presumably someone is compiling against glibc 2.38 headers and then trying to link against an older glibc version.

**Comment by nickdesaulniers** [comment]
Based on https://android.googlesource.com/kernel/build/+/refs/heads/main/kleaf/docs/debugging.md and a comment in the souces:
```
# Debug flag to control make verbosity
# E (default) = Error (-s)
# I = Info
# D = Debug (V=1)
# V = Verbose (V=2)
```
, it sounds like kleaf has `--debug_make_verbosity=D` can be set to enable `V=1`. @Aarqw12 can you please rebuild with that set and redirect all output to a file, then attach that here (or post a link to a gist)?

**Comment by nathanchance** [comment]
I can reproduce this using AOSP's LLVM (so it does not appear to be a recent regression nor one strictly related to my kernel.org builds) in a Clear Linux container and also an Arch Linux one (which also has glibc 2.38), so I don't think this is distribution specific... I cannot reproduce it in a Debian Sid container, which has glibc 2.36, so...

> Feedback from #glibc on IRC
> 
> > Presumably someone is compiling against glibc 2.38 headers and then trying to link against an older glibc version.

This seems to be what is happening if I understand correctly. `libsubcmd-in.o` is linked from all the `tools/bpf/resolve_btfids/libsubcmd` objects, which appears to be built using the system headers (as I do not see any `-I` or `-include` in the command from outside the kernel build tree). However, when `libsubcmd-in.o` is linked into `resolve_btfids` (through `libsubcmd.a`), it appears that there are a bunch of `-L` flags to make the build more hermetic; Bazel makes them pretty unreadable but I can only assume this means that the final link happens against Android's older glibc?

The full build log is available at https://gist.github.com/nathanchance/0ceb37caf0ac2c227f3d361a6d1aa3a4.

**Comment by jacky8hyf** [comment]
A quick parsing of the "last line" in https://gist.github.com/nathanchance/0ceb37caf0ac2c227f3d361a6d1aa3a4 gives me this:

```shell
clang \
  <out_dir>/common/tools/bpf/resolve_btfids//resolve_btfids-in.o\
 -L<...>libbase.so<...>\
 -L<...>libc++.so<...>\
 -L<...>libcrypto-host.so<...>\
 -L<...>libcrypto.so<...>\
 -L<...>libcrypto_Uutils.so<...>\
 -L<...>libcutils.so<...>\
 -L<...>libelf.so<...>\
 -L<...>libext2_Ublkid-host.so<...>\
 -L<...>libext2_Ucom_Uerr-host.so<...>\
 -L<...>libext2_Ue2p-host.so<...>\
 -L<...>libext2_Uquota-host.so<...>\
 -L<...>libext2_Uuuid-host.so<...>\
 -L<...>libext2fs-host.so<...>\
 -L<...>libext4_Uutils.so<...>\
 -L<...>libfdt.so<...>\
 -L<...>libicui18n-host.so<...>\
 -L<...>libicuuc-host.so<...>\
 -L<...>libinterceptor.so<...>\
 -L<...>liblog.so<...>\
 -L<...>liblp.so<...>\
 -L<...>libsparse-host.so<...>\
 -L<...>libsqlite.so<...>\
 -L<...>libz-host.so<...>\
  -fuse-ld=lld --rtlib=compiler-rt \
  --sysroot=<exec_root>/build/kernel/build-tools/sysroot \
  -o <out_dir>/common/tools/bpf/resolve_btfids//resolve_btfids \
  <out_dir>/common/tools/bpf/resolve_btfids//libbpf/libbpf.a \
  <out_dir>/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a \
  -lelf -lz
```

**Comment by jacky8hyf** [comment]
In the build log @nathanchance provided, here's how libsubcmd-in.o is built:

```shell
   ld.lld -r -o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/libsubcmd-in.o  <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/exec-cmd.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/help.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/pager.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/parse-options.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/run-command.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/sigchain.o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/subcmd-config.o
```

... which aligns with what @nathanchance says in 

> libsubcmd-in.o is linked from all the tools/bpf/resolve_btfids/libsubcmd objects

Then, for example, here's how `exec-cmd.o` was built:

```
  clang -Wp,-MD,<out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/.exec-cmd.o.d -Wp,-MT,<out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/exec-cmd.o -ggdb3 -Wall -Wextra -std=gnu99 -fPIC -O3 -Werror -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -I<root_dir>/common/tools/include/ -Wbad-function-cast -Wdeclaration-after-statement -Wformat-security -Wformat-y2k -Winit-self -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wno-system-headers -Wold-style-definition -Wpacked -Wredundant-decls -Wstrict-prototypes -Wswitch-default -Wswitch-enum -Wundef -Wwrite-strings -Wformat -Wno-type-limits -Wshadow -D"BUILD_STR(s)=#s" -c -o <out_dir>/common/tools/bpf/resolve_btfids/libsubcmd/exec-cmd.o exec-cmd.c
```

The only `-I` is `common/tools/include/`. 

This appears to align with HOSTCFLAGS / USERCFLAGS prior to the toolchain resolution change (the `--incompatible_kernel_use_resolved_toolchains` flag), unless I am missing something?

https://cs.android.com/android/kernel/superproject/+/common-android-mainline:build/kernel/_setup_env.sh;l=130;drc=82f2c2f0f75758ff2d8baaaec4e96bc48a641640

**Comment by jacky8hyf** [comment]
> Bazel makes them pretty unreadable but I can only assume this means that the final link happens against Android's older glibc?

The "final link" appears to be links against linux_x86_imported_libs, defined here:

https://cs.android.com/android/kernel/superproject/+/common-android-mainline:prebuilts/kernel-build-tools/BUILD.bazel;l=18;drc=0dcade4d8f3d6215d99e19fe9d3753df90a10dbe

which are these libs:

https://cs.android.com/android/kernel/superproject/+/common-android-mainline:prebuilts/kernel-build-tools/linux-x86/lib64/

I am confused; are these objects host (linux-x86_64) or device (android-arm64)?

**Comment by nathanchance** [comment]
> ```shell
> clang \
>   <out_dir>/common/tools/bpf/resolve_btfids//resolve_btfids-in.o\
>  -L<...>libbase.so<...>\
>  -L<...>libc++.so<...>\
>  -L<...>libcrypto-host.so<...>\
>  -L<...>libcrypto.so<...>\
>  -L<...>libcrypto_Uutils.so<...>\
>  -L<...>libcutils.so<...>\
>  -L<...>libelf.so<...>\
>  -L<...>libext2_Ublkid-host.so<...>\
>  -L<...>libext2_Ucom_Uerr-host.so<...>\
>  -L<...>libext2_Ue2p-host.so<...>\
>  -L<...>libext2_Uquota-host.so<...>\
>  -L<...>libext2_Uuuid-host.so<...>\
>  -L<...>libext2fs-host.so<...>\
>  -L<...>libext4_Uutils.so<...>\
>  -L<...>libfdt.so<...>\
>  -L<...>libicui18n-host.so<...>\
>  -L<...>libicuuc-host.so<...>\
>  -L<...>libinterceptor.so<...>\
>  -L<...>liblog.so<...>\
>  -L<...>liblp.so<...>\
>  -L<...>libsparse-host.so<...>\
>  -L<...>libsqlite.so<...>\
>  -L<...>libz-host.so<...>\
>   -fuse-ld=lld --rtlib=compiler-rt \
>   --sysroot=<exec_root>/build/kernel/build-tools/sysroot \
>   -o <out_dir>/common/tools/bpf/resolve_btfids//resolve_btfids \
>   <out_dir>/common/tools/bpf/resolve_btfids//libbpf/libbpf.a \
>   <out_dir>/common/tools/bpf/resolve_btfids//libsubcmd/libsubcmd.a \
>   -lelf -lz
> ```

@jacky8hyf Thanks for actually formatting it out! I am now able to see the `--sysroot` argument, which reveals:

```
$ ls -al build/kernel/build-tools/sysroot/usr/lib/libc.so.6
lrwxrwxrwx 1 nathan nathan 12 Sep  6 17:50 build/kernel/build-tools/sysroot/usr/lib/libc.so.6 -> libc-2.17.so
```

which confirms the original theory... However, in `build/kernel` and my build log, I can see an `export HOSTCFLAGS` with the `--sysroot` and `-I` pointing to the `glibc-2.17` sysroot. It appears that prior to commit [13e07691a16f](https://git.kernel.org/linus/13e07691a16ff31b209fbfce25c01ff296b05e45) ("tools/resolve_btfids: Alter how HOSTCC is forced"), `HOSTCFLAGS` was not being taken into account when building `libsubcmd` for `resolve_btfids`... I think this was missed by commit [0e3a1c902ffb](https://git.kernel.org/linus/0e3a1c902ffb56e9fe4416f0cd382c97b09ecbf6) ("tools/resolve_btfids: Build with host flags").

All that is to say that this diff against `android14-6.1` appears to resolve the issue for me:

```diff
diff --git a/tools/bpf/resolve_btfids/Makefile b/tools/bpf/resolve_btfids/Makefile
index 19a3112e271a..60afe9432cb1 100644
--- a/tools/bpf/resolve_btfids/Makefile
+++ b/tools/bpf/resolve_btfids/Makefile
@@ -49,7 +49,7 @@ $(OUTPUT) $(OUTPUT)/libsubcmd $(LIBBPF_OUT):
        $(Q)mkdir -p $(@)

 $(SUBCMDOBJ): fixdep FORCE | $(OUTPUT)/libsubcmd
-       $(Q)$(MAKE) -C $(SUBCMD_SRC) OUTPUT=$(abspath $(dir $@))/ $(abspath $@)
+       $(Q)$(MAKE) -C $(SUBCMD_SRC) EXTRA_CFLAGS="$(CFLAGS)" OUTPUT=$(abspath $(dir $@))/ $(abspath $@)

 $(BPFOBJ): $(wildcard $(LIBBPF_SRC)/*.[ch] $(LIBBPF_SRC)/Makefile) | $(LIBBPF_OUT)
        $(Q)$(MAKE) $(submake_extras) -C $(LIBBPF_SRC) OUTPUT=$(LIBBPF_OUT)    \
```

We could look into backporting 13e07691a16f to `linux-6.1.y` but I recall a couple of follow up fixes that were needed for it so I am not sure that is the safest option (plus I have not even looked into if there are any conflicts). I think that diff should probably just be applied to Android.

> I am confused; are these objects host (linux-x86_64) or device (android-arm64)?

These are object files for host tools.

**Comment by jacky8hyf** [comment]
If I understand correct, you have already find a proper way forward, and you don't need to build with flags like `--nonoincompatible_kernel_use_resolved_toolchains`, is that right?

**Comment by nathanchance** [comment]
> If I understand correct, you have already find a proper way forward, and you don't need to build with flags like `--nonoincompatible_kernel_use_resolved_toolchains`, is that right?

I think so, I submitted https://android-review.googlesource.com/c/kernel/common/+/2745896.

**Comment by nathanchance** [comment]
Greg KH requested the proposed patch be submitted to stable upstream and Ian suggested looking into backporting the upstream solution into 6.1 instead, which I did: https://lore.kernel.org/20230908161526.GA3344687@dev-arch.thelio-3990X/

**Comment by Aarqw12** [comment]
> Greg KH requested the proposed patch be submitted to stable upstream and Ian suggested looking into backporting the upstream solution into 6.1 instead, which I did: https://lore.kernel.org/20230908161526.GA3344687@dev-arch.thelio-3990X/

I see in kernel 5.10 also its affected for linkage failure using : BUILD_CONFIG=private/gs-google/build.config.slider build/build.sh  

I cherry pick this commit https://android-review.googlesource.com/c/kernel/common/+/2745896
its return this error : 
[error.txt](https://github.com/ClangBuiltLinux/linux/files/12564103/error.txt)


in fix temporary I added -Wno-gcc-compat

$(SUBCMDOBJ): fixdep FORCE | $(OUTPUT)/libsubcmd
	$(Q)$(MAKE) -C $(SUBCMD_SRC) EXTRA_CFLAGS="$(CFLAGS) -Wno-gcc-compat" OUTPUT=$(abspath $(dir $@))/ $(abspath $@)

**Comment by Aarqw12** [comment]
> > Greg KH requested the proposed patch be submitted to stable upstream and Ian suggested looking into backporting the upstream solution into 6.1 instead, which I did: https://lore.kernel.org/20230908161526.GA3344687@dev-arch.thelio-3990X/
> 
> I see in kernel 5.10 also its affected for linkage failure using : BUILD_CONFIG=private/gs-google/build.config.slider build/build.sh
> 
> I cherry pick this commit https://android-review.googlesource.com/c/kernel/common/+/2745896 its return this error : [error.txt](https://github.com/ClangBuiltLinux/linux/files/12564103/error.txt)
> 
> in fix temporary I added -Wno-gcc-compat
> 
> (SUBCMDOBJ):fixdepFORCE|(OUTPUT)/libsubcmd (Q)(MAKE) -C (SUBCMDSRC)EXTRACFLAGS="(CFLAGS) -Wno-gcc-compat" OUTPUT=$(abspath (dir@))/ (abspath@)

I backported in kernel 5.10 https://github.com/torvalds/linux/commit/50560ce6a0bdab2fc37384c52aa02c7043909d2c and compile is fixed without -Wno-gcc-compat

**Comment by nathanchance** [comment]
I don't think `-Wno-gcc-compat` is necessary, fixing that warning is quite trivial.

I have submitted the following series to ACK to hopefully resolve this issue once and for all.

5.10
- https://android-review.googlesource.com/c/kernel/common/+/2746220
- https://android-review.googlesource.com/c/kernel/common/+/2746221

5.15
- https://android-review.googlesource.com/c/kernel/common/+/2747613
- https://android-review.googlesource.com/c/kernel/common/+/2747614

6.1
- https://android-review.googlesource.com/c/kernel/common/+/2744841
- https://android-review.googlesource.com/c/kernel/common/+/2744842
- https://android-review.googlesource.com/c/kernel/common/+/2744843
- https://android-review.googlesource.com/c/kernel/common/+/2744844
- https://android-review.googlesource.com/c/kernel/common/+/2744845
- https://android-review.googlesource.com/c/kernel/common/+/2744846
- https://android-review.googlesource.com/c/kernel/common/+/2744847
- https://android-review.googlesource.com/c/kernel/common/+/2744848
- https://android-review.googlesource.com/c/kernel/common/+/2744849
- https://android-review.googlesource.com/c/kernel/common/+/2744850

5.10 and 5.15 cannot be fixed in stable because this issue should not even be visible without commit [0e3a1c902ffb](https://git.kernel.org/linus/0e3a1c902ffb56e9fe4416f0cd382c97b09ecbf6) ("tools/resolve_btfids: Build with host flags"), which landed in 5.18.

**Comment by Aarqw12** [comment]
> I don't think `-Wno-gcc-compat` is necessary, fixing that warning is quite trivial.
> 
> I have submitted the following series to ACK to hopefully resolve this issue once and for all.
> 
> 5.10
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2746220
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2746221
> 
> 
> 5.15
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2747613
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2747614
> 
> 
> 6.1
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744841
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744842
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744843
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744844
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744845
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744846
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744847
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744848
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744849
> 
>     * https://android-review.googlesource.com/c/kernel/common/+/2744850
> 
> 
> 5.10 and 5.15 cannot be fixed in stable because this issue should not even be visible without commit [0e3a1c902ffb](https://git.kernel.org/linus/0e3a1c902ffb56e9fe4416f0cd382c97b09ecbf6) ("tools/resolve_btfids: Build with host flags"), which landed in 5.18.

thanks I confirm in 5.10 & 6.1 its compile fine in full LTO and build boot normally ( using latest slimm LLVM 17.0.0-rc4 )

**Comment by nickdesaulniers** [comment]
@nathanchance 's request for stable to pick up fixes for 6.1 have been accepted.
https://lore.kernel.org/all/20230908161526.GA3344687@dev-arch.thelio-3990X/

As such, I've abandoned the android14-6.1 patches from android review.

I've updated the rest to be rebased and reference an internal bug I've filed with more info.

Finally, I cherry picked the android14-5.15 patches to android13-5.15.

**Comment by nickdesaulniers** [comment]
I've gotten all of the remaining patches merged up into AOSP.

@Aarqw12 thanks for the report.

@nathanchance thanks for tracking down the fixes after figuring out a reproducer.

I think we're all set here now.
